### PR TITLE
Superdense coding kata finish

### PR DESCRIPTION
**Pull request by Manvi-Agrawal**
Part of #1185.

Changes Done:
- Revamped Task 1:  Using `qAlice` and `qBob` for better clarity. Replaced `\rangle` with `\ket`
- Added tasks 2-4.
- Using a tuple of boolean variables instead of `ProtocolMessage` to improve learner and developer productivity.

### CODE REVIEW

File: katas/content/superdense_coding/Common.qs [hunk 1]
```diff
@@ -0,0 +1,67 @@
+namespace Kata.Verification {
+    open Kata;
```
Comment (Manvi-Agrawal) [inline katas/content/superdense_coding/Common.qs:2]: Oddly enough, removing this gives me : `main.qs@(1,1): type error: type ProtocolMessage cannot be converted into a string`

File: katas/content/superdense_coding/alice_sends_message/Solution.qs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+namespace Kata {
+    newtype ProtocolMessage = (Bit1 : Bool, Bit2 : Bool);
+
```
Comment (Manvi-Agrawal) [inline katas/content/superdense_coding/alice_sends_message/Solution.qs:3]:
```suggestion
```

File: katas/content/index.json [hunk 3]
```diff
@@ -9,6 +9,7 @@
   "preparing_states",
   "single_qubit_measurements",
   "multi_qubit_measurements",
+  "superdense_coding",
   "random_numbers",
```
Comment (tcNickolas) [inline katas/content/index.json:13]: Let's move this one after QRNG, that's usually how these applications are taught, and that matches the order in the classic katas. (Teleportation will be between these two)
```suggestion
  "random_numbers",
  "superdense_coding",
```

File: katas/content/superdense_coding/Common.qs (hunk 1 above)
Comment (tcNickolas) [reply katas/content/superdense_coding/Common.qs:2]: Makes sense, the namespaces right now are not hierarchical, so Kata.Verification doesn't have access to Kata namespace. (That's why we call solutions with Kata. prefix)

File: katas/content/superdense_coding/alice_sends_message/Placeholder.qs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+namespace Kata {
+    newtype ProtocolMessage = (Bit1 : Bool, Bit2 : Bool);
+
+    operation EncodeMessageInQubit(qAlice : Qubit, message : ProtocolMessage) : Unit {
```
Comment (tcNickolas) [inline katas/content/superdense_coding/alice_sends_message/Placeholder.qs:4]: Looking at this kata in the modern system, I think we should drop ProtocolMessage type altogether, and rewrite it as just a tuple of two bits, like this:
```suggestion
    operation EncodeMessageInQubit(qAlice : Qubit, (bit1, bit2) : (Int, Int)) : Unit {
```

It will make the code much shorter and more straightforward, for example, we won't have to teach bit access in the kata. And in this task it makes things confusing, actually, since we're asking them to implement solution within an if that is supposed to teach bit access, but the solution also needs to include another, independent if outside of this one... Plus then we can drop protocol notes in task description

**Review by tcNickolas — CHANGES_REQUESTED** [review]
I took a quick look, things are looking good overall, but I think we need to get rid of `ProtocolMessage` and just use tuples. The custom types don't really work great with new infrastructure, with us having to redefine them each time, and since it's not structured as a multi-file Q# project, IntelliSense fails to recognize them too... I think using tuples will improve the experience both for the learner and for developer.

I'll take a more detailed pass once you remove `ProtocolMessage`, it will be a lot easier at that point.

Thank you!

File: katas/content/index.json (hunk 3 above)
Comment (Manvi-Agrawal) [reply katas/content/index.json]: Done. Thanks for your suggestion.

File: katas/content/superdense_coding/alice_sends_message/Placeholder.qs (hunk 4 above)
Comment (Manvi-Agrawal) [reply katas/content/superdense_coding/alice_sends_message/Placeholder.qs]: Reverted to tuple of Booleans by checking out to previous commit.

File: katas/content/superdense_coding/index.md [hunk 5]
```diff
@@ -18,10 +18,41 @@ We split the superdense coding protocol into several steps:
...
+    "path": "bob_decodes_message",
+    "qsDependencies": [
+        "../KatasLibrary.qs",
+        "./Common.qs"
+    ]
+})
+
+@[exercise]({
+    "id": "superdense_coding__protocol_e2e",
+    "title": "Superdense coding protocol: End to End",
```
Comment (Manvi-Agrawal) [inline katas/content/superdense_coding/index.md:52]:
```suggestion
    "title": "Superdense Coding Protocol: End to End",
```

File: katas/content/superdense_coding/index.md [hunk 6]
```diff
@@ -18,10 +18,41 @@ We split the superdense coding protocol into several steps:
...
+    "path": "alice_sends_message",
+    "qsDependencies": [
+        "../KatasLibrary.qs",
+        "./Common.qs"
+    ]
+})
+
+@[exercise]({
+    "id": "superdense_coding__bob_decodes_message",
+    "title": "Bob's task: Decode the Message and Reset the qubits",
```
Comment (Manvi-Agrawal) [inline katas/content/superdense_coding/index.md:42]:
```suggestion
    "title": "Bob's task: Decode the Message and Reset the Qubits",
```

**Comment by Manvi-Agrawal** [comment]
> I'll take a more detailed pass once you remove ProtocolMessage, it will be a lot easier at that point.
@tcNickolas , could you please take a look? I have removed `ProtocolMessage` class. Thank you.

File: katas/content/superdense_coding/alice_sends_message/solution.md [hunk 7]
```diff
@@ -0,0 +1,33 @@
+Recall that we learnt how to prepare all Bell states in "Preparing Quantum States" kata. This is slightly advanced version of that task demonstrating the "power of entanglement", where operations applied to only one qubit affect the state of the system in interesting way.
+
+Let's use the below Bell state we prepared in the previous task:
+
+$$\frac{1}{\sqrt{2}} \big(\ket{00} + \ket{11} \big)$$
+
+We can transform it according to the value of bits:
```
Comment (tcNickolas) [inline katas/content/superdense_coding/alice_sends_message/solution.md:7]:
```suggestion
We can transform it into every other Bell state according to the value of `message`:
```

File: katas/content/superdense_coding/alice_sends_message/solution.md [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+- `(0, 0)`: $\ket{\Phi^{+}} = \frac{1}{\sqrt{2}} (\ket{00} + \ket{11})$
```
Comment (tcNickolas) [inline katas/content/superdense_coding/alice_sends_message/solution.md:9]: These two lists are rendered as one 8-point list, which looks a bit odd. I would merge the pairs of bullet points that describe the same scenario into one point each, so that each value of message is described and handled right away

File: katas/content/superdense_coding/alice_sends_message/solution.md [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+- `(0, 1)`: $\ket{\Psi^{+}} = \frac{1}{\sqrt{2}} (\ket{01} + \ket{10})$
+- `(1, 0)`: $\ket{\Phi^{-}} = \frac{1}{\sqrt{2}} (\ket{00} - \ket{11})$
+- `(1, 1)`: $\ket{\Psi^{-}} = \frac{1}{\sqrt{2}} (\ket{01} - \ket{10})$
+
+- If `bits == (0, 0)`, we do nothing - the prepared state is already $\ket{\Phi^{+}}$.
+
+- If `bits ==  (0, 1)`, we need to change the second qubit in both $\ket{00}$ and $\ket{11}$ terms. Observe that applying an $X$ gate to Alice's qubit does exactly that:
+  $$X \otimes I (\ket{\Phi^{+}}) = \frac{1}{\sqrt{2}} (\ket{10} + \ket{01})$$
```
Comment (tcNickolas) [inline katas/content/superdense_coding/alice_sends_message/solution.md:17]: Here and in the next formula we need brackets around (X \otimes I) rather than around the \ket for Bell state (just like you did in the last formula) - \ket is already a single entity, but getting a two-qubit gate out of the tensor product needs emphasizing

File: katas/content/superdense_coding/alice_sends_message/solution.md [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+- If `bits == (1, 0)`, we need to add a relative phase of $-1$ to the $\ket{11}$ term. Observe that applying $Z$ gate to Alice's qubit does exactly that:
+  $$Z \otimes I (\ket{\Phi^{+}}) = \frac{1}{\sqrt{2}} (\ket{00} - \ket{11})$$
+
+- If `bits = (1, 1)`, we use the same logic to realize that we need to apply both the $Z$ and $X$ corrections to get $\ket{\Psi^{-}}$ state.
+  $$ (Z \otimes I).(X \otimes I) (\ket{\Psi^{+}}) = (Z \otimes I) \frac{1}{\sqrt{2}} (\ket{10} + \ket{01}) = \frac{1}{\sqrt{2}} (-\ket{10} + \ket{01}) = \frac{1}{\sqrt{2}} (\ket{01} - \ket{10}) $$
```
Comment (tcNickolas) [inline katas/content/superdense_coding/alice_sends_message/solution.md:23]:
```suggestion
  $$ (Z \otimes I) \cdot (X \otimes I) \ket{\Psi^{+}} = (Z \otimes I) \frac{1}{\sqrt{2}} (\ket{10} + \ket{01}) = \frac{1}{\sqrt{2}} (-\ket{10} + \ket{01}) = \frac{1}{\sqrt{2}} (\ket{01} - \ket{10}) $$
```

File: katas/content/superdense_coding/alice_sends_message/solution.md [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+The final sequence of steps is as follows:
+
+1. Apply the $X$ gate to the alice's qubit if `bit2 == 1`.
+2. Apply the $Z$ gate to the alice's qubit if `bit1 == 1`.
```
Comment (tcNickolas) [inline katas/content/superdense_coding/alice_sends_message/solution.md:28]:
```suggestion
1. Apply the $X$ gate to Alice's qubit if `bit2 == 1`.
2. Apply the $Z$ gate to Alice's qubit if `bit1 == 1`.
```

File: katas/content/superdense_coding/bob_decodes_message/Placeholder.qs [hunk 12]
```diff
@@ -0,0 +1,14 @@
+namespace Kata {
+    operation DecodeMessageFromQubits(qAlice : Qubit, qBob : Qubit) : (Bool, Bool) {
+        // Implement your solution here...
+
+        // Returning the qubit states after measuring
+        return (MResetZ(qAlice) == One, MResetZ(qBob) == One);
+    }
+
+    // You might find this helper operation from an earlier task useful.
+    operation CreateEntangledPair(qAlice : Qubit, qBob : Qubit) : Unit is Adj {
```
Comment (tcNickolas) [inline katas/content/superdense_coding/bob_decodes_message/Placeholder.qs:10]: I thought we were going to drop this helper from this task and assume that the learner will just type in CNOT and H?

File: katas/content/superdense_coding/bob_decodes_message/Placeholder.qs [hunk 13]
```diff
@@ -0,0 +1,14 @@
+namespace Kata {
+    operation DecodeMessageFromQubits(qAlice : Qubit, qBob : Qubit) : (Bool, Bool) {
+        // Implement your solution here...
+
+        // Returning the qubit states after measuring
+        return (MResetZ(qAlice) == One, MResetZ(qBob) == One);
```
Comment (tcNickolas) [inline katas/content/superdense_coding/bob_decodes_message/Placeholder.qs:6]: I would do a placeholder with `return (false, false)`, and let the learner do the rest - this way we're giving away most of the solution already :-)

File: katas/content/superdense_coding/bob_decodes_message/Solution.qs [hunk 14]
```diff
@@ -0,0 +1,12 @@
+namespace Kata {
+    operation DecodeMessageFromQubits(qAlice : Qubit, qBob : Qubit) : (Bool, Bool) {
+        Adjoint CreateEntangledPair(qAlice, qBob);
```
Comment (tcNickolas) [inline katas/content/superdense_coding/bob_decodes_message/Solution.qs:3]: Same here, we can type CNOT and H and drop CreateEntangledPair altogether

File: katas/content/superdense_coding/bob_decodes_message/solution.md [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+
+- `(0, 0)`: $\ket{\Phi^{+}} = \frac{1}{\sqrt{2}} (\ket{00} + \ket{11})$
+- `(0, 1)`: $\ket{\Psi^{+}} = \frac{1}{\sqrt{2}} (\ket{01} + \ket{10})$
+- `(1, 0)`: $\ket{\Phi^{-}} = \frac{1}{\sqrt{2}} (\ket{00} - \ket{11})$
+- `(1, 1)`: $\ket{\Psi^{-}} = \frac{1}{\sqrt{2}} (\ket{01} - \ket{10})$
+
+To get our state back, we can undo the entanglement by applying $CNOT$ and $H$ gate.
+Notice that it's important to keep the order right. The qubits that are subject to the Hadamard transform and the $CNOT$ gate in the preparation of the pair have to match the operations below, or the order of the data bits will get flipped.
+
+Notice that [`Adjoint`](https://learn.microsoft.com/azure/quantum/user-guide/language/expressions/functorapplication#adjoint-functor) functor in Q# does exactly that.
```
Comment (tcNickolas) [inline katas/content/superdense_coding/bob_decodes_message/solution.md:11]: This is confusing to me - we're looking to get a message rather than a state, and it's unclear Adjoint of what operation we're discussing. I'd rephrase as something like this:

```suggestion
To read out the encoded message, Bob needs to figure out which of the four Bell states he has. We can map the Bell states to basis states by applying a $CNOT$ gate with the first qubit as control and the second qubit as target, followed by an $H$ gate on the first qubit. (Notice that this is exactly what [`Adjoint`](https://learn.microsoft.com/azure/quantum/user-guide/language/expressions/functorapplication#adjoint-functor) of the state preparation operation in the first task does.)
```

File: katas/content/superdense_coding/protocol_e2e/Placeholder.qs [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+
+    // You might find this helper operation from an earlier task useful.
+    operation CreateEntangledPair(qAlice : Qubit, qBob : Qubit) : Unit is Adj {
+        H(qAlice);
+        CNOT(qAlice, qBob);
+    }
+
+    // You might find this helper operation from an earlier task useful.
+    operation EncodeMessageInQubit(qAlice : Qubit, message : (Bool, Bool)) : Unit {
+        let (cbit1, cbit2) = message;
```
Comment (tcNickolas) [inline katas/content/superdense_coding/protocol_e2e/Placeholder.qs:16]: Same suggestion cbit -> bit here

File: katas/content/superdense_coding/protocol_e2e/Placeholder.qs [hunk 17]
```diff
@@ -0,0 +1,33 @@
+namespace Kata {
+    operation SuperdenseCodingProtocol(message : (Bool, Bool)) : (Bool, Bool) {
+        // Implement your solution here...
+
+        return (false, false);
+    }
+
+    // You might find this helper operation from an earlier task useful.
```
Comment (tcNickolas) [inline katas/content/superdense_coding/protocol_e2e/Placeholder.qs:8]: I'd use plural here
```suggestion
    // You might find these helper operations from earlier tasks useful.
```
And then we can remove the same comment from the next two tasks, to make them more compact

File: katas/content/superdense_coding/protocol_e2e/index.md [hunk 18]
```diff
@@ -0,0 +1,9 @@
+**Input:** Two classical bits to be transmitted.
+
+**Goal:**
```
Comment (tcNickolas) [inline katas/content/superdense_coding/protocol_e2e/index.md:3]:
```suggestion
**Goal:**
```

File: katas/content/superdense_coding/protocol_e2e/index.md [hunk 19]
```diff
@@ -0,0 +1,9 @@
+**Input:** Two classical bits to be transmitted.
+
+**Goal:**  
+Put together the steps performed in previous tasks to implement the full superdense coding protocol:
+
+- Prepare an EPR Pair
```
Comment (tcNickolas) [inline katas/content/superdense_coding/protocol_e2e/index.md:6]: We've not used this term here before, so let's stick to terms we used
```suggestion
- Prepare a Bell state.
```

File: katas/content/superdense_coding/protocol_e2e/index.md [hunk 20]
```diff
@@ -0,0 +1,9 @@
...
+- Encode the two classical bits in the state of the pair by applying quantum gates to one member of the pair,
```
Comment (tcNickolas) [inline katas/content/superdense_coding/protocol_e2e/index.md:7]:
```suggestion
- Encode the two classical bits in the state of the qubits by applying quantum gates to one qubit of the Bell state,
```

File: katas/content/superdense_coding/protocol_e2e/index.md [hunk 21]
```diff
@@ -0,0 +1,9 @@
...
+- Decode the two classical bits from the state of the pair.
```
Comment (tcNickolas) [inline katas/content/superdense_coding/protocol_e2e/index.md:8]:
```suggestion
- Decode the two classical bits from the resulting state.
```

File: katas/content/superdense_coding/protocol_e2e/solution.md [hunk 22]
```diff
@@ -0,0 +1,11 @@
+We can put together the operations defined in the previous tasks to run superdense protocol E2E.
+
+- Prepare an EPR Pair
```
Comment (tcNickolas) [inline katas/content/superdense_coding/protocol_e2e/solution.md:3]: Same update here to match the task description

File: katas/content/superdense_coding/protocol_e2e/solution.md [hunk 23]
```diff
@@ -0,0 +1,11 @@
+We can put together the operations defined in the previous tasks to run superdense protocol E2E.
```
Comment (tcNickolas) [inline katas/content/superdense_coding/protocol_e2e/solution.md:1]:
```suggestion
We can put together the operations defined in the previous tasks to run superdense coding protocol end-to-end.
```

**Review by tcNickolas — CHANGES_REQUESTED** [review]
Looks great! I know I left a lot of comments, but they're mostly polish to keep things nice and uniform. One last round of edits, and this is ready to ship :-)

Thank you!

File: katas/content/superdense_coding/alice_sends_message/Placeholder.qs [hunk 24]
```diff
@@ -0,0 +1,14 @@
...
+
+    operation EncodeMessageInQubit(qAlice : Qubit, message : (Bool, Bool)) : Unit {
+        // Get the bits from the message
+        let (bit1, bit2) = message;
+
+        // Accessing bit1
+        if bit1 {
+            // Implement your solution here...
+
+        }
```
[the remark this answers is not in the document]
Comment (Manvi-Agrawal) [reply katas/content/superdense_coding/alice_sends_message/Placeholder.qs:11]: Done, I wanted to do it this way since its cleaner.

I dont know if something changed. Earlier, it was complaining about some return type (Bool, Bool) issue and adding if block suppresed that error :-)

File: katas/content/superdense_coding/bob_decodes_message/Placeholder.qs (hunk 12 above)
Comment (Manvi-Agrawal) [reply katas/content/superdense_coding/bob_decodes_message/Placeholder.qs:10]: I thought of keeping it for the convenience of the learner, in case they might  want to use it. I was more of favor of dropping it here, in case we drop the last End-To-End task, which requires us to copy-paste all the 3 solutions. 

@tcNickolas , Kindly let me know if you think, this should be dropped here. In that case, I was thinking to update the solution text to avoid mention of `Adjoint`  functor.

File: katas/content/superdense_coding/alice_sends_message/solution.md (hunk 9 above)
Comment (Manvi-Agrawal) [reply katas/content/superdense_coding/alice_sends_message/solution.md]: Done

File: katas/content/superdense_coding/alice_sends_message/solution.md (hunk 8 above)
Comment (Manvi-Agrawal) [reply katas/content/superdense_coding/alice_sends_message/solution.md:9]: Nice catch. Updated solution.

File: katas/content/superdense_coding/index.md [hunk 25]
```diff
@@ -18,10 +18,41 @@ We split the superdense coding protocol into several steps:
 - Finally, we compose those steps into the complete superdense coding protocol.
 
 @[exercise]({
```
[the remark this answers is not in the document]
Comment (Manvi-Agrawal) [reply katas/content/superdense_coding/index.md:20]: Done

File: katas/content/superdense_coding/alice_sends_message/Solution.qs [hunk 26]
```diff
@@ -0,0 +1,15 @@
+namespace Kata {
+
+    operation EncodeMessageInQubit(qAlice : Qubit, message : (Bool, Bool)) : Unit {
+        // Get the bits from the message
+        let (cbit1, cbit2) = message;
```
[the remark this answers is not in the document]
Comment (Manvi-Agrawal) [reply katas/content/superdense_coding/alice_sends_message/Solution.qs]: Done, initially I wanted to call them cbit in placeholder and solution to denote that they are classical bits. And it explaination I was bit1 with supplementary info in text that said classical bits, etc. It seems that might confuse the learner, so using `bit1` everywhere

File: katas/content/superdense_coding/bob_decodes_message/Placeholder.qs (hunk 13 above)
Comment (Manvi-Agrawal) [reply katas/content/superdense_coding/bob_decodes_message/Placeholder.qs]: Thanks for the suggestion. I updated the testing harness to do a reset of the qubits.

I was getting the below error without it, so I used reset operation to get away with that. 

```text
[ERR_ASSERTION]: Exercise "superdense_coding__bob_decodes_message" has compilation or runtime errors
        //  when using the placeholder as solution. Compilation and runtime errors:
        //   Message = runtime error: Qubit1 released while not in |0⟩ state
       ```

**Comment by Manvi-Agrawal** [comment]
> Looks great! I know I left a lot of comments, but they're mostly polish to keep things nice and uniform. One last round of edits, and this is ready to ship :-)
@tcNickolas , thanks for your review. I have addressed most of your comments. Could you please take a look again?

File: katas/content/superdense_coding/index.md [hunk 27]
```diff
@@ -18,10 +18,41 @@ We split the superdense coding protocol into several steps:
...
+    "path": "alice_sends_message",
+    "qsDependencies": [
+        "../KatasLibrary.qs",
+        "./Common.qs"
+    ]
+})
+
+@[exercise]({
+    "id": "superdense_coding__bob_decodes_message",
+    "title": "Bob's task: Decode the Message and Reset the Qubits",
```
[the remark this answers is not in the document]
Comment (Manvi-Agrawal) [reply katas/content/superdense_coding/index.md]: Changed. This was in the previous kata:-)

File: katas/content/superdense_coding/bob_decodes_message/Placeholder.qs (hunk 12 above)
Comment (tcNickolas) [reply katas/content/superdense_coding/bob_decodes_message/Placeholder.qs:9]: I would drop it here (but keep it in the e2e task, since you're right that we need all 3 solutions in it anyways). We can mention the Adjoint in the solution text even if we're not using it in the code - it's a useful bit of knowledge that using adjoint of state prep allows us to do measurement in Bell state even if we choose to implement adjoint by hand

Comment (Manvi-Agrawal) [reply katas/content/superdense_coding/bob_decodes_message/Placeholder.qs]: @tcNickolas, done.

**Comment by Manvi-Agrawal** [comment]
@tcNickolas , thanks for your review. I have addressed all your comments. Could you please take a look again?

File: katas/content/superdense_coding/alice_sends_message/index.md [hunk 28]
```diff
@@ -0,0 +1,18 @@
...
+<details>
+  <summary><b>Need a hint?</b></summary>
+Manipulate Alice's half of the entangled pair to change the joint state of the two qubits to one of the following four states based on the value of message:
+
+- `(0, 0)`: $\ket{\Phi^{+}} = \frac{1}{\sqrt{2}} (\ket{00} + \ket{11})$
+- `(0, 1)`: $\ket{\Psi^{+}} = \frac{1}{\sqrt{2}} (\ket{01} + \ket{10})$
+- `(1, 0)`: $\ket{\Phi^{-}} = \frac{1}{\sqrt{2}} (\ket{00} - \ket{11})$
+- `(1, 1)`: $\ket{\Psi^{-}} = \frac{1}{\sqrt{2}} (\ket{01} - \ket{10})$
+
+</details>
```
Comment (tcNickolas) [inline katas/content/superdense_coding/alice_sends_message/index.md:18]: I just realized that we don't actually explain here how we test the solution. In all the previous katas, if the outcome of a task was expected to be a state, like here, the results would be a state comparison, but here for some reason we're talking about transfer. (And this was a problem in the classic kata too, I just never looked at it with a fresh eye! Revisiting everything as part of this migration is not a bad thing :-))

I suggest we rephrase a bit, to give more context on testing and to move expected encoding from hint to main task - strictly speaking, it's possible to implement superdense coding with a different mapping of messages to Bell states, it's just the customary approach:
```suggestion
Superdense coding protocol changes the joint state of the two qubits to one of the following four states based on the value of message:

- `(0, 0)`: $\ket{\Phi^{+}} = \frac{1}{\sqrt{2}} (\ket{00} + \ket{11})$
- `(0, 1)`: $\ket{\Psi^{+}} = \frac{1}{\sqrt{2}} (\ket{01} + \ket{10})$
- `(1, 0)`: $\ket{\Phi^{-}} = \frac{1}{\sqrt{2}} (\ket{00} - \ket{11})$
- `(1, 1)`: $\ket{\Psi^{-}} = \frac{1}{\sqrt{2}} (\ket{01} - \ket{10})$

Note that your solution is tested as part of an end-to-end implementation of the protocol; the goal is to get the message transmitted correctly.
```

File: katas/content/superdense_coding/alice_sends_message/solution.md [hunk 29]
```diff
@@ -0,0 +1,35 @@
...
+$$\frac{1}{\sqrt{2}} \big(\ket{00} + \ket{11} \big)$$
+
+We can transform it into every other Bell state according to the value of `message`:
+
+- `(0, 0)`: $\ket{\Phi^{+}} = \frac{1}{\sqrt{2}} (\ket{00} + \ket{11})$
+- `(0, 1)`: $\ket{\Psi^{+}} = \frac{1}{\sqrt{2}} (\ket{01} + \ket{10})$
+- `(1, 0)`: $\ket{\Phi^{-}} = \frac{1}{\sqrt{2}} (\ket{00} - \ket{11})$
+- `(1, 1)`: $\ket{\Psi^{-}} = \frac{1}{\sqrt{2}} (\ket{01} - \ket{10})$
+
+### Detailed Explaination
```
Comment (tcNickolas) [inline katas/content/superdense_coding/alice_sends_message/solution.md:14]: We use subheaders as headers, when there's a different solution or a different topic raised. Here, the goal of the solution is already to provide a detailed explanation, so I'd rephrase
```suggestion
Here is how we can perform this transformation:
```

File: katas/content/superdense_coding/bob_decodes_message/index.md [hunk 30]
```diff
@@ -0,0 +1,6 @@
+**Inputs:**
+
+1. `qAlice` : Qubit received from Alice.
+2. `qBob` : Bob's part of the entangled pair.
+
+**Goal** :  Decode the message using the qubit received from Alice and reset both qubits to a $\ket{00}$ state. For this, retrieve two bits of classic data from the qubits and return a tuple of the form `(Bool, Bool)` to represent `message`. The state of the qubits in the end of the operation should be $\ket{00}$.
```
Comment (tcNickolas) [inline katas/content/superdense_coding/bob_decodes_message/index.md:6]: We need a note on testing here as well. Also, since we've modified the testing harness, there is no requirement to reset qubits to 0, the solution passes without that

```suggestion
**Goal** :  Decode the message using the qubit received from Alice. For this, retrieve two bits of classic data from the qubits and return a tuple of the form `(Bool, Bool)` to represent `message`. The state of the qubits in the end of the operation does not matter.

Note that your solution is tested as part of an end-to-end implementation of the protocol; the goal is to get the message transmitted correctly.
```

File: katas/content/superdense_coding/Common.qs [hunk 31]
```diff
@@ -0,0 +1,79 @@
...
+        CreateEntangledPairWrapper_Reference([qAlice, qBob]);
+
+        encodeOp(qAlice, message);
+
+        let (bit1, bit2) = decodeOp(qAlice, qBob);
+
+        // We get this error if dont call Reset
+        // [ERR_ASSERTION]: Exercise "superdense_coding__bob_decodes_message" has compilation or runtime errors
+        //  when using the placeholder as solution. Compilation and runtime errors:
+        //   Message = runtime error: Qubit1 released while not in |0⟩ state
```
Comment (tcNickolas) [inline katas/content/superdense_coding/Common.qs:47]: You don't have to explain the Reset. It's a known behavior that qubits have to be in 0 state before release, you'll notice that, for example, in state preparation tasks, if the state prepared is not what we want (and thus after adjoint is applied the state is not 0), we reset the state before returning false, without much emphasis really:
```suggestion
```

**Review by tcNickolas — APPROVED** [review]
Looks great! Of course, I had a realization that we're not explaining why the messages from testing harness look the way they do, so I think we need to add notes about that to two tasks... But that's the final round of suggestions, promise!

Thank you!

File: katas/content/superdense_coding/index.md [hunk 32]
```diff
@@ -18,10 +18,38 @@ We split the superdense coding protocol into several steps:
...
+})
+
+@[exercise]({
+    "id": "superdense_coding__protocol_e2e",
+    "title": "Superdense Coding Protocol: End to End",
+    "path": "protocol_e2e",
+    "qsDependencies": [
+        "./Common.qs"
     ]
 })
```
Comment (tcNickolas) [inline katas/content/superdense_coding/index.md:55]: And another thing I just realized - we try to add a little conclusion to the end of the katas even if they didn't have them before
```suggestion
})


@[section]({
    "id": "superdense_coding__conclusion",
    "title": "Conclusion"
})

Congratulations! In this kata you learned to implement superdense coding protocol.
```

File: katas/content/superdense_coding/Common.qs (hunk 31 above)
Comment (Manvi-Agrawal) [reply katas/content/superdense_coding/Common.qs:47]: I thought of adding here in case someone ends up getting confused of the difference from original harness. But probably we can omit that, since its us developers only who would be referring old katas for short time :-)

File: katas/content/superdense_coding/alice_sends_message/index.md (hunk 28 above)
Comment (Manvi-Agrawal) [reply katas/content/superdense_coding/alice_sends_message/index.md:18]: Nice catch, Done :)

File: katas/content/superdense_coding/index.md [hunk 33]
```diff
@@ -18,10 +18,46 @@ We split the superdense coding protocol into several steps:
...
+
+@[exercise]({
+    "id": "superdense_coding__protocol_e2e",
+    "title": "Superdense Coding Protocol: End to End",
+    "path": "protocol_e2e",
+    "qsDependencies": [
+        "./Common.qs"
+    ]
+})
+
```
Comment (Manvi-Agrawal) [inline katas/content/superdense_coding/index.md:56]:
```suggestion
```

**Review by tcNickolas — APPROVED** [review]
Looks great, thank you!
